include/rocksdb/env_encrypt2.h

```diff
@@ -0,0 +1,471 @@
+// nounce (assumed to be packed/byte aligned)
+typedef struct {
+  uint8_t key_description_[EVP_MAX_MD_SIZE];
+  uint8_t nonce_[AES_BLOCK_SIZE / 2];  // block size is 16
+} Prefix0_t;
+
+class AESBlockAccessCipherStream : public BlockAccessCipherStream {
+ public:
+  AESBlockAccessCipherStream(const AesCtrKey_t& key, uint8_t code_version,
+                             uint8_t nonce[])
+      : key_(key), code_version_(code_version) {
+    memcpy(&nonce_, nonce, AES_BLOCK_SIZE / 2);
+  }
+
+  // BlockSize returns the size of each block supported by this cipher stream.
+  virtual size_t BlockSize() override { return AES_BLOCK_SIZE; };
+
+ protected:
+  // Allocate scratch space which is passed to EncryptBlock/DecryptBlock.
+  virtual void AllocateScratch(std::string&) override{};
+
+  // Encrypt a block of data at the given block index.
+  // Length of data is equal to BlockSize();
+  virtual Status EncryptBlock(uint64_t blockIndex, char* data,
+                              char* scratch) override;
+
+  // Decrypt a block of data at the given block index.
+  // Length of data is equal to BlockSize();
+  virtual Status DecryptBlock(uint64_t blockIndex, char* data,
+                              char* scratch) override;
+
+  AesCtrKey_t key_;
+  uint8_t code_version_;
+  uint8_t nonce_[AES_BLOCK_SIZE / 2];
+};
+
+class CTREncryptionProvider2 : public EncryptionProvider {
+ public:
+  CTREncryptionProvider2() = delete;
+
+  CTREncryptionProvider2(const CTREncryptionProvider&&) = delete;
+
+  CTREncryptionProvider2(const Sha1Description_t& key_desc,
+                         const AesCtrKey_t& key)
+      : valid_(false), key_desc_(key_desc), key_(key) {
+    valid_ = key_desc_.IsValid() && key_.IsValid();
+  }
+
+  CTREncryptionProvider2(const std::string& key_desc_str,
+                         const uint8_t unformatted_key[], int bytes)
+      : valid_(false), key_desc_(key_desc_str), key_(unformatted_key, bytes) {
+    valid_ = key_desc_.IsValid() && key_.IsValid();
+  }
+
+  virtual size_t GetPrefixLength() override {
+    return sizeof(Prefix0_t) + sizeof(EncryptMarker_t);
+  }
+
+  virtual Status CreateNewPrefix(const std::string& /*fname*/, char* prefix,
+                                 size_t prefixLength) override;
+
+  virtual Status CreateCipherStream(
+      const std::string& /*fname*/, const EnvOptions& /*options*/,
+      Slice& /*prefix*/,
+      std::unique_ptr<BlockAccessCipherStream>* /*result*/) override {
+    return Status::NotSupported("Wrong EncryptionProvider assumed");
+  }
+
+  virtual BlockAccessCipherStream* CreateCipherStream2(uint8_t code_version,
+                                                       uint8_t nonce[]) {
+    return new AESBlockAccessCipherStream(key_, code_version, nonce);
+  }
+
+  bool Valid() const { return valid_; };
+  const Sha1Description_t& key_desc() const { return key_desc_; };
+  const AesCtrKey_t& key() const { return key_; };
```

Comment:
We hit this all the time with gcc 4.8, which is why we added the build. Ctor param shadows this member function.

Reply:
@pdillinger Thank you for taking time to explain.  I can easily adapt the constructor.

Reply:
this is complete and passing tests now.